File: src/nzbAvailability.ts
import { connect as tlsConnect, TLSSocket } from "tls";
import { connect as netConnect, Socket } from "net";
import { createHash } from "crypto";
import { Item } from "./types.js";
import { getCache, setCache } from "./cache.js";

interface AvailabilityResult {
  ok: boolean;
  reason?: string;
  /** Fraction of sampled articles available across all configured servers */
  coverage?: number;
}

const CACHE_TTL = 86400; // 24h - article retention is stable
const NZB_FETCH_TIMEOUT = 5000;
const NNTP_CONNECT_TIMEOUT = 8000;
const NNTP_CMD_TIMEOUT = 8000;
const MAX_NZB_SIZE = 5 * 1024 * 1024;

/**
 * Coverage threshold: a release passes if at least this fraction of its
 * RAR-part first segments exist on at least one configured server.
 * 1.0 = strictest (all parts must be available); the streaming server
 * needs every part to assemble a valid stream, so anything below 100%
 * means probable mid-stream failure.
 */
const REQUIRED_COVERAGE = 1.0;

interface NntpServer {
  protocol: "nntp" | "nntps";
  host: string;
  port: number;
  user: string;
  pass: string;
}

const parseNntpUrl = (url: string): NntpServer | null => {
  try {
    const u = new URL(url);
    const protocol = u.protocol === "nntps:" ? "nntps" : "nntp";
    if (!u.hostname || !u.port) return null;
    return {
      protocol,
      host: u.hostname,
      port: parseInt(u.port, 10),
      user: decodeURIComponent(u.username),
      pass: decodeURIComponent(u.password),
    };
  } catch {
    return null;
  }
};

const sha1Short = (s: string) =>
  createHash("sha1").update(s).digest("hex").slice(0, 16);

/**
 * Pick a small set of canary segments that, if available, are very
 * likely to mean the whole release plays. Block-account backbones
 * frequently retain article HEADERS while the BODY data is gone, so we
 * deliberately probe the *body* of:
 *   1. first segment of part01.rar (RAR archive header)
 *   2. last segment of part01.rar (RAR header CRC, which Stremio's NZB
 *      engine reads first via getFileSize - this is what fails most
 *      often when bodies are lost)
 *   3. last segment of the highest-numbered .partNN.rar (end of media)
 *
 * Three samples gives high signal on whether the release is actually
 * playable, without paying for every part.
 */
const extractCanaryMessageIds = (xml: string): string[] => {
  const partFiles: Array<{ part: number; first: string; last: string }> = [];
  const fileRe =
    /<file\b[^>]*subject="[^"]*\.part(\d+)\.rar[^"]*"[^>]*>([\s\S]*?)<\/file>/gi;

  for (const m of xml.matchAll(fileRe)) {
    const part = parseInt(m[1], 10);
    const segments: Array<{ n: number; id: string }> = [];
    for (const sm of m[2].matchAll(
      /<segment\b[^>]*\bnumber="(\d+)"[^>]*>([^<]+)<\/segment>/gi,
    )) {
      segments.push({ n: parseInt(sm[1], 10), id: sm[2].trim() });
    }
    if (!segments.length) continue;
    segments.sort((a, b) => a.n - b.n);
    partFiles.push({
      part,
      first: segments[0].id,
      last: segments[segments.length - 1].id,
    });
  }

  if (!partFiles.length) return [];

  partFiles.sort((a, b) => a.part - b.part);
  const part01 = partFiles[0];
  const partLast = partFiles[partFiles.length - 1];

  // De-duplicate: a single-part release reduces to 2 message-ids
  const ids = new Set<string>([part01.first, part01.last, partLast.last]);
  return Array.from(ids);
};

/**
 * Probe articles via NNTP BODY. We ignore the body data itself - as soon
 * as the response code arrives (222 = body follows, 430 = no such
 * article) we know the answer and tear down the connection. This avoids
 * downloading hundreds of MB while still using the only NNTP command
 * that actually checks for body availability.
 *
 * STAT/HEAD return 223/221 even for articles whose bodies have been
 * purged on block-account backbones (UsenetExpress, etc), so they're
 * useless for this check.
 */
const probeArticlesOnServer = async (
  server: NntpServer,
  messageIds: string[],
): Promise<boolean[]> => {
  return new Promise((resolve) => {
    const results: boolean[] = new Array(messageIds.length).fill(false);
    const socket: TLSSocket | Socket =
      server.protocol === "nntps"
        ? tlsConnect({
            host: server.host,
            port: server.port,
            servername: server.host,
            rejectUnauthorized: false,
          })
        : netConnect({ host: server.host, port: server.port });

    socket.setEncoding("utf8");
    socket.setTimeout(NNTP_CONNECT_TIMEOUT);

    let stage:
      | "greet"
      | "auth-user"
      | "auth-pass"
      | "body"
      | "done" = "greet";
    let cursor = 0;
    let buf = "";

    const finish = () => {
      stage = "done";
      try {
        socket.write("QUIT\r\n");
      } catch {
        // ignore
      }
      socket.destroy();
      resolve(results);
    };

    const fail = () => {
      socket.destroy();
      resolve(results);
    };

    socket.on("timeout", fail);
    socket.on("error", fail);

    const sendNextOrFinish = () => {
      if (cursor < messageIds.length) {
        socket.write(`BODY <${messageIds[cursor]}>\r\n`);
      } else {
        finish();
      }
    };

    socket.on("data", (chunk: string) => {
      buf += chunk;
      while (true) {
        const eol = buf.indexOf("\r\n");
        if (eol < 0) break;
        const line = buf.slice(0, eol);
        buf = buf.slice(eol + 2);

        const code = parseInt(line.slice(0, 3), 10);
        if (stage === "done") continue;

        switch (stage) {
          case "greet":
            if (code === 200 || code === 201) {
              stage = "auth-user";
              socket.setTimeout(NNTP_CMD_TIMEOUT);
              socket.write(`AUTHINFO USER ${server.user}\r\n`);
            } else {
              return fail();
            }
            break;
          case "auth-user":
            if (code === 381) {
              stage = "auth-pass";
              socket.write(`AUTHINFO PASS ${server.pass}\r\n`);
            } else if (code === 281) {
              stage = "body";
              if (!messageIds.length) return finish();
              sendNextOrFinish();
            } else {
              return fail();
            }
            break;
          case "auth-pass":
            if (code === 281) {
              stage = "body";
              if (!messageIds.length) return finish();
              sendNextOrFinish();
            } else {
              return fail();
            }
            break;
          case "body":
            if (code === 222) {
              // Body would now stream; mark as available and reset the
              // connection. Reusing the same socket while a body is
              // mid-flight risks data corruption, so we tear down and
              // start fresh for the next article.
              results[cursor] = true;
              cursor++;
              socket.destroy();
              if (cursor < messageIds.length) {
                // Continue with a fresh connection
                probeArticlesOnServer(server, messageIds.slice(cursor)).then(
                  (rest) => {
                    rest.forEach((ok, i) => (results[cursor + i] = ok));
                    resolve(results);
                  },
                );
              } else {
                resolve(results);
              }
              return;
            } else {
              // 430 / anything else = not available
              results[cursor] = false;
              cursor++;
              sendNextOrFinish();
            }
            break;
          default:
            break;
        }
      }
    });
  });
};

const probeAvailability = async (
  nzbUrl: string,
  servers: NntpServer[],
): Promise<AvailabilityResult> => {
  let xml: string;
  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), NZB_FETCH_TIMEOUT);
    const response = await fetch(nzbUrl, { signal: ctrl.signal });
    clearTimeout(t);
    if (!response.ok) return { ok: false, reason: `nzb-http-${response.status}` };
    xml = await response.text();
    if (xml.length > MAX_NZB_SIZE) return { ok: false, reason: "nzb-too-large" };
  } catch (err) {
    return {
      ok: false,
      reason: err instanceof Error ? err.message : "nzb-fetch-failed",
    };
  }

  const messageIds = extractCanaryMessageIds(xml);
  if (!messageIds.length) return { ok: false, reason: "no-rar-parts" };

  // Probe every server in parallel. We require an article to be on EVERY
  // configured server (AND, not OR) because Stremio's NZB engine doesn't
  // reliably fall back between backbones - if eunews lacks an article,
  // it 500s with "not on any backbones" even when bonus has the bytes.
  // Strict behaviour matches what the streaming server actually needs.
  const perServer = await Promise.all(
    servers.map((s) => probeArticlesOnServer(s, messageIds)),
  );

  const covered = messageIds.map((_, idx) =>
    perServer.every((r) => r[idx] === true),
  );
  const coverage = covered.filter(Boolean).length / messageIds.length;

  if (coverage >= REQUIRED_COVERAGE) return { ok: true, coverage };

  const missing = messageIds.length - covered.filter(Boolean).length;
  return {
    ok: false,
    reason: `articles-missing-${missing}-of-${messageIds.length}`,
    coverage,
  };
};

export const checkNzbAvailability = async (
  nzbUrl: string,
  serverUrls: string[],
): Promise<AvailabilityResult> => {
  const servers = serverUrls
    .map(parseNntpUrl)
    .filter((s): s is NntpServer => s !== null);
  if (!servers.length) return { ok: false, reason: "no-valid-servers" };

  const cacheKey = `nzb-avail:${sha1Short(
    nzbUrl + "|" + serverUrls.slice().sort().join(","),
  )}`;
  const cached = await getCache<AvailabilityResult>(cacheKey);
  if (cached) return cached;

  const result = await probeAvailability(nzbUrl, servers);
  await setCache(cacheKey, result, CACHE_TTL);
  return result;
};

export const filterByNzbAvailability = async (
  items: Item[],
  serverUrls: string[],
  getNzbUrl: (item: Item) => string,
): Promise<Item[]> => {
  if (!items.length || !serverUrls.length) return items;

  const checks = await Promise.all(
    items.map(async (item) => ({
      item,
      avail: await checkNzbAvailability(getNzbUrl(item), serverUrls),
    })),
  );

  const dropped = checks.filter((c) => !c.avail.ok);
  if (dropped.length) {
    console.log(
      `[nzbAvailability] excluded ${dropped.length} of ${items.length}: ${dropped
        .slice(0, 3)
        .map(
          (c) =>
            `"${c.item.title}" (${c.avail.reason}${
              c.avail.coverage !== undefined
                ? `, ${Math.round(c.avail.coverage * 100)}% coverage`
                : ""
            })`,
        )
        .join(", ")}${dropped.length > 3 ? "..." : ""}`,
    );
  }

  return checks.filter((c) => c.avail.ok).map((c) => c.item);
};

File: src-rust/nzb_availability.rs
//! NNTP `BODY`-probe of canary message-IDs to verify an NZB is actually
//! playable on the user's configured backbones.
//!
//! Mirrors `src/nzbAvailability.ts`. Three canary articles are probed per NZB:
//!   1. First segment of `part01.rar`
//!   2. Last segment of `part01.rar`  ← critical (Stremio's `getFileSize` reads it)
//!   3. Last segment of the highest `partNN.rar`
//!
//! Coverage is **strict AND across servers** — every server must have every
//! canary article, matching Stremio's NZB engine's real fall-back behavior.

use anyhow::{anyhow, Result};
use nzb_rs::{File, Nzb};
use once_cell::sync::Lazy;
use regex::Regex;
use rustls::ClientConfig;
use serde::{Deserialize, Serialize};
use sha1::{Digest, Sha1};
use std::sync::Arc;
use std::time::Duration;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::time::timeout;
use tokio_rustls::TlsConnector;
use url::Url;

use crate::cache::AVAILABILITY_CACHE;
use crate::nzb_api::{item_nzb_url, Item};

const FETCH_TIMEOUT: Duration = Duration::from_secs(5);
const MAX_NZB_SIZE: u64 = 5 * 1024 * 1024;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AvailabilityResult {
    pub ok: bool,
    pub reason: Option<String>,
}

fn sha1_short(s: &str) -> String {
    let digest = Sha1::digest(s.as_bytes());
    hex::encode(digest)[..16].to_string()
}

/// `partNN.rar` (case-insensitive) — captures `NN` for finding "highest part".
static PART_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?i)\.part(\d+)\.rar(?:[^.\w]|$)").expect("part regex"));

fn part_number(file_name: &str) -> Option<u32> {
    PART_RE
        .captures(file_name)
        .and_then(|c| c.get(1)?.as_str().parse::<u32>().ok())
}

/// Pick (first_part01_seg, last_part01_seg, last_partNN_seg) message-IDs.
/// Returns `None` if the NZB has no `partNN.rar` files at all.
///
/// **Sort by `seg.number` before picking first/last** so the canary set
/// matches what `streaming::preflight` actually fetches. NZB XML *usually*
/// lists segments in number order but the spec doesn't require it; if the
/// two paths disagree on which article is "first", availability could
/// pass on one segment while pre-flight dies on a different one.
fn pick_canaries(nzb: &Nzb) -> Option<Vec<&str>> {
    let parts: Vec<(u32, &File)> = nzb
        .files
        .iter()
        .filter_map(|f| f.name().and_then(|n| part_number(n).map(|p| (p, f))))
        .collect();

    if parts.is_empty() {
        return None;
    }

    let part01 = parts.iter().min_by_key(|(p, _)| *p).map(|(_, f)| *f)?;
    let highest = parts.iter().max_by_key(|(p, _)| *p).map(|(_, f)| *f)?;

    let part01_sorted: Vec<&nzb_rs::Segment> = sorted_segments(part01);
    let highest_sorted: Vec<&nzb_rs::Segment> = sorted_segments(highest);

    let part01_first = part01_sorted.first().copied()?;
    let part01_last = part01_sorted.last().copied()?;
    let highest_last = highest_sorted.last().copied()?;

    let mut canaries: Vec<&str> = vec![
        part01_first.message_id.as_str(),
        part01_last.message_id.as_str(),
    ];

    // Avoid probing the same article twice if part01 == highest part.
    if !std::ptr::eq(part01, highest) {
        canaries.push(highest_last.message_id.as_str());
    }

    Some(canaries)
}

/// Borrow `file.segments` sorted by `seg.number`. Mirror of the sort
/// `streaming::preflight::probe_rar_inner` does so canary picks line up
/// with what pre-flight actually fetches.
fn sorted_segments(file: &File) -> Vec<&nzb_rs::Segment> {
    let mut v: Vec<&nzb_rs::Segment> = file.segments.iter().collect();
    v.sort_by_key(|s| s.number);
    v
}

/// Build the AVAILABILITY_CACHE key for a given (nzb_url, server_urls) pair.
/// Public so external callers (notably `streaming::preflight`) can invalidate
/// stale entries without re-implementing the hash.
pub fn cache_key_for(nzb_url: &str, server_urls: &[String]) -> String {
    let mut sorted = server_urls.to_vec();
    sorted.sort();
    format!(
        "nzb-avail:{}",
        sha1_short(&format!("{nzb_url}|{}", sorted.join(",")))
    )
}

/// Drop any cached "ok" verdict for this NZB. Called by `streaming::preflight`
/// when the smoke segment turns out to be missing on every server — the
/// cached verdict is stale (article aged out / DMCA'd between the original
/// probe and now) and shouldn't be trusted on the next search.
pub async fn invalidate_for(nzb_url: &str, server_urls: &[String]) {
    let key = cache_key_for(nzb_url, server_urls);
    AVAILABILITY_CACHE.invalidate(&key).await;
    tracing::info!("[nzbAvailability] invalidated stale cache entry for nzb_url");
}

pub async fn check_nzb_availability(
    client: &reqwest::Client,
    server_urls: &[String],
    nzb_url: &str,
) -> AvailabilityResult {
    let cache_key = cache_key_for(nzb_url, server_urls);

    if let Some(hit) = AVAILABILITY_CACHE.get(&cache_key).await {
        return hit;
    }

    let result = probe(client, server_urls, nzb_url).await;
    AVAILABILITY_CACHE.insert(cache_key, result.clone()).await;
    result
}

async fn probe(
    client: &reqwest::Client,
    server_urls: &[String],
    nzb_url: &str,
) -> AvailabilityResult {
    let resp = match client.get(nzb_url).timeout(FETCH_TIMEOUT).send().await {
        Ok(r) => r,
        Err(err) => {
            return AvailabilityResult {
                ok: false,
                reason: Some(format!(
                    "fetch-failed: {}",
                    crate::util::redact_log(&err.to_string())
                )),
            };
        }
    };

    if !resp.status().is_success() {
        return AvailabilityResult {
            ok: false,
            reason: Some(format!("http-{}", resp.status().as_u16())),
        };
    }

    if let Some(cl) = resp.content_length() {
        if cl > MAX_NZB_SIZE {
            return AvailabilityResult {
                ok: false,
                reason: Some(format!("nzb-too-large-{cl}")),
            };
        }
    }

    let xml = match resp.text().await {
        Ok(t) => t,
        Err(err) => {
            return AvailabilityResult {
                ok: false,
                reason: Some(format!(
                    "read-failed: {}",
                    crate::util::redact_log(&err.to_string())
                )),
            };
        }
    };

    if (xml.len() as u64) > MAX_NZB_SIZE {
        return AvailabilityResult {
            ok: false,
            reason: Some(format!("nzb-too-large-{}", xml.len())),
        };
    }

    let nzb = match Nzb::parse(&xml) {
        Ok(n) => n,
        Err(err) => {
            return AvailabilityResult {
                ok: false,
                reason: Some(format!(
                    "parse-failed: {}",
                    crate::util::redact_log(&err.to_string())
                )),
            };
        }
    };

    let canaries = match pick_canaries(&nzb) {
        Some(c) => c,
        None => {
            // No RAR parts — fall back to the largest non-par2 file's first
            // and last segments. Matches Usenet-Ultimate's "obfuscated
            // filename" path; gives availability checks a chance to work on
            // single-file releases too.
            let largest = nzb
                .files
                .iter()
                .filter(|f| !f.is_par2())
                .max_by_key(|f| f.size());
            let Some(f) = largest else {
                return AvailabilityResult {
                    ok: false,
                    reason: Some("no-files".to_string()),
                };
            };
            let first = match f.segments.first() {
                Some(s) => s.message_id.as_str(),
                None => {
                    return AvailabilityResult {
                        ok: false,
                        reason: Some("no-segments".to_string()),
                    };
                }
            };
            let last = f.segments.last().map(|s| s.message_id.as_str());
            let mut v = vec![first];
            if let Some(l) = last {
                if l != first {
                    v.push(l);
                }
            }
            v
        }
    };

    // Strict AND coverage: every server probes every canary.
    // Probes for one server are sequential (cleanest reuse of body_probe's
    // dedicated socket per article); servers are checked sequentially too,
    // since failure on any server short-circuits.
    for server in server_urls {
        let safe_server = crate::util::redact_url(server);
        for canary in &canaries {
            match body_probe(server, canary).await {
                Ok(true) => continue,
                Ok(false) => {
                    return AvailabilityResult {
                        ok: false,
                        reason: Some(format!("missing-on-server: {safe_server}")),
                    };
                }
                Err(err) => {
                    return AvailabilityResult {
                        ok: false,
                        reason: Some(format!(
                            "probe-error on {safe_server}: {}",
                            crate::util::redact_log(&err.to_string())
                        )),
                    };
                }
            }
        }
    }

    AvailabilityResult {
        ok: true,
        reason: None,
    }
}

/// Filter items by NZB availability. Drops items that fail any canary probe
/// on any configured server. Runs probes in parallel across items.
pub async fn filter_by_nzb_availability(
    client: &reqwest::Client,
    items: Vec<Item>,
    server_urls: &[String],
) -> Vec<Item> {
    if items.is_empty() || server_urls.is_empty() {
        return items;
    }

    let total = items.len();
    let checks = items.into_iter().map(|item| {
        let client = client.clone();
        let servers = server_urls.to_vec();
        async move {
            let url = item_nzb_url(&item);
            let res = check_nzb_availability(&client, &servers, &url).await;
            (item, res)
        }
    });
    let results: Vec<(Item, AvailabilityResult)> = futures::future::join_all(checks).await;

    let dropped: Vec<&(Item, AvailabilityResult)> = results.iter().filter(|(_, r)| !r.ok).collect();

    if !dropped.is_empty() {
        let preview: Vec<String> = dropped
            .iter()
            .take(20)
            .map(|(it, r)| format!("\"{}\" ({})", it.title, r.reason.as_deref().unwrap_or("?")))
            .collect();
        let suffix = if dropped.len() > 3 { "..." } else { "" };
        tracing::info!(
            "[nzbAvailability] excluded {} of {}: {}{}",
            dropped.len(),
            total,
            preview.join(", "),
            suffix
        );
    }

    results
        .into_iter()
        .filter(|(_, r)| r.ok)
        .map(|(it, _)| it)
        .collect()
}

/// NNTP `BODY` probe of a single message-id. Opens a fresh connection
/// each time (no pooling — probes fire infrequently and we want each
/// check to reflect the current connection state of the server). 5s
/// timeout. Returns `Ok(true)` on `222`, `Ok(false)` on `430/423`, error
/// on auth/network failure.
async fn body_probe(server_url: &str, message_id: &str) -> Result<bool> {
    timeout(
        Duration::from_secs(5),
        body_probe_inner(server_url, message_id),
    )
    .await
    .map_err(|_| anyhow!("Connection timeout (5s)"))?
}

async fn body_probe_inner(server_url: &str, message_id: &str) -> Result<bool> {
    let url = Url::parse(server_url)?;
    let secure = url.scheme() == "nntps";
    let host = url
        .host_str()
        .ok_or_else(|| anyhow!("missing host"))?
        .to_string();
    let port = url.port().unwrap_or(if secure { 563 } else { 119 });
    let username = urlencoding::decode(url.username())?.into_owned();
    let password = urlencoding::decode(url.password().unwrap_or(""))?.into_owned();

    let tcp = TcpStream::connect((host.as_str(), port)).await?;

    if secure {
        let mut roots = rustls::RootCertStore::empty();
        roots.extend(webpki_roots::TLS_SERVER_ROOTS.iter().cloned());
        let mut tls_cfg = ClientConfig::builder()
            .with_root_certificates(roots)
            .with_no_client_auth();
        tls_cfg
            .dangerous()
            .set_certificate_verifier(Arc::new(NoVerify));
        let connector = TlsConnector::from(Arc::new(tls_cfg));
        let server_name = rustls::pki_types::ServerName::try_from(host.clone())
            .map_err(|_| anyhow!("invalid server name"))?;
        let stream = connector.connect(server_name, tcp).await?;
        run_body_probe(stream, &username, &password, message_id).await
    } else {
        run_body_probe(tcp, &username, &password, message_id).await
    }
}

async fn run_body_probe<S>(
    mut stream: S,
    username: &str,
    password: &str,
    message_id: &str,
) -> Result<bool>
where
    S: AsyncReadExt + AsyncWriteExt + Unpin,
{
    let mut buf = String::new();
    let mut tmp = [0u8; 1024];
    let mut authenticated = username.is_empty() && password.is_empty();

    loop {
        let n = stream.read(&mut tmp).await?;
        if n == 0 {
            return Err(anyhow!("Connection closed unexpectedly"));
        }
        buf.push_str(std::str::from_utf8(&tmp[..n])?);

        while let Some(idx) = buf.find("\r\n") {
            let line = buf[..idx].to_string();
            buf.drain(..idx + 2);

            let code: u16 = line.get(..3).and_then(|s| s.parse().ok()).unwrap_or(0);

            if (code == 200 || code == 201) && !authenticated {
                if !username.is_empty() && !password.is_empty() {
                    stream
                        .write_all(format!("AUTHINFO USER {username}\r\n").as_bytes())
                        .await?;
                } else {
                    authenticated = true;
                    stream
                        .write_all(format!("BODY <{message_id}>\r\n").as_bytes())
                        .await?;
                }
            } else if code == 381 {
                stream
                    .write_all(format!("AUTHINFO PASS {password}\r\n").as_bytes())
                    .await?;
            } else if code == 281 {
                authenticated = true;
                stream
                    .write_all(format!("BODY <{message_id}>\r\n").as_bytes())
                    .await?;
            } else if (480..490).contains(&code) {
                return Err(anyhow!("Authentication failed: {line}"));
            } else if code == 222 {
                // Article exists; drop socket without reading body.
                return Ok(true);
            } else if code == 430 || code == 423 {
                // No such article (430) / no article with that number (423).
                return Ok(false);
            } else if code >= 500 {
                return Err(anyhow!("Server error: {line}"));
            }
        }
    }
}

#[derive(Debug)]
struct NoVerify;

impl rustls::client::danger::ServerCertVerifier for NoVerify {
    fn verify_server_cert(
        &self,
        _end_entity: &rustls::pki_types::CertificateDer<'_>,
        _intermediates: &[rustls::pki_types::CertificateDer<'_>],
        _server_name: &rustls::pki_types::ServerName<'_>,
        _ocsp_response: &[u8],
        _now: rustls::pki_types::UnixTime,
    ) -> std::result::Result<rustls::client::danger::ServerCertVerified, rustls::Error> {
        Ok(rustls::client::danger::ServerCertVerified::assertion())
    }

    fn verify_tls12_signature(
        &self,
        _message: &[u8],
        _cert: &rustls::pki_types::CertificateDer<'_>,
        _dss: &rustls::DigitallySignedStruct,
    ) -> std::result::Result<rustls::client::danger::HandshakeSignatureValid, rustls::Error> {
        Ok(rustls::client::danger::HandshakeSignatureValid::assertion())
    }

    fn verify_tls13_signature(
        &self,
        _message: &[u8],
        _cert: &rustls::pki_types::CertificateDer<'_>,
        _dss: &rustls::DigitallySignedStruct,
    ) -> std::result::Result<rustls::client::danger::HandshakeSignatureValid, rustls::Error> {
        Ok(rustls::client::danger::HandshakeSignatureValid::assertion())
    }

    fn supported_verify_schemes(&self) -> Vec<rustls::SignatureScheme> {
        use rustls::SignatureScheme::*;
        vec![
            RSA_PKCS1_SHA256,
            RSA_PKCS1_SHA384,
            RSA_PKCS1_SHA512,
            ECDSA_NISTP256_SHA256,
            ECDSA_NISTP384_SHA384,
            ECDSA_NISTP521_SHA512,
            RSA_PSS_SHA256,
            RSA_PSS_SHA384,
            RSA_PSS_SHA512,
            ED25519,
        ]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn part_number_extracts_correctly() {
        assert_eq!(part_number("Movie.part01.rar"), Some(1));
        assert_eq!(part_number("Movie.part001.rar yEnc"), Some(1));
        assert_eq!(part_number("Movie.part42.rar (1/10)"), Some(42));
        assert_eq!(part_number("Movie.PART15.rar"), Some(15)); // case-insensitive
        assert_eq!(part_number("Movie.r00"), None);
        assert_eq!(part_number("Movie.rar"), None);
        assert_eq!(part_number("Movie.mkv"), None);
    }

    #[test]
    fn pick_canaries_with_multiple_parts() {
        let xml = r#"<?xml version="1.0" encoding="UTF-8"?>
<nzb xmlns="http://www.newzbin.com/DTD/2003/nzb">
  <file poster="x" date="1700000000" subject="Show.part01.rar yEnc (1/3) 12345">
    <groups><group>alt.binaries.test</group></groups>
    <segments>
      <segment bytes="100" number="1">part01-seg1@x</segment>
      <segment bytes="100" number="2">part01-seg2@x</segment>
      <segment bytes="100" number="3">part01-seg3@x</segment>
    </segments>
  </file>
  <file poster="x" date="1700000000" subject="Show.part02.rar yEnc (1/2) 12345">
    <groups><group>alt.binaries.test</group></groups>
    <segments>
      <segment bytes="100" number="1">part02-seg1@x</segment>
      <segment bytes="100" number="2">part02-seg2@x</segment>
    </segments>
  </file>
</nzb>"#;
        let nzb = Nzb::parse(xml).expect("valid nzb");
        let canaries = pick_canaries(&nzb).expect("has rars");
        assert_eq!(canaries.len(), 3);
        // first segment of part01
        assert_eq!(canaries[0], "part01-seg1@x");
        // last segment of part01 — critical: Stremio's getFileSize reads this
        assert_eq!(canaries[1], "part01-seg3@x");
        // last segment of highest partNN (part02)
        assert_eq!(canaries[2], "part02-seg2@x");
    }

    #[test]
    fn pick_canaries_dedupes_when_only_one_part() {
        let xml = r#"<?xml version="1.0" encoding="UTF-8"?>
<nzb xmlns="http://www.newzbin.com/DTD/2003/nzb">
  <file poster="x" date="1700000000" subject="Show.part01.rar yEnc (1/2) 12345">
    <groups><group>alt.binaries.test</group></groups>
    <segments>
      <segment bytes="100" number="1">single-seg1@x</segment>
      <segment bytes="100" number="2">single-seg2@x</segment>
    </segments>
  </file>
</nzb>"#;
        let nzb = Nzb::parse(xml).expect("valid nzb");
        let canaries = pick_canaries(&nzb).expect("has rar");
        // Should not include part02 entry since highest == part01.
        assert_eq!(canaries.len(), 2);
        assert_eq!(canaries[0], "single-seg1@x");
        assert_eq!(canaries[1], "single-seg2@x");
    }

    #[test]
    fn pick_canaries_returns_none_for_no_rars() {
        let xml = r#"<?xml version="1.0" encoding="UTF-8"?>
<nzb xmlns="http://www.newzbin.com/DTD/2003/nzb">
  <file poster="x" date="1700000000" subject="Movie.mkv yEnc (1/1) 12345">
    <groups><group>alt.binaries.test</group></groups>
    <segments>
      <segment bytes="100" number="1">mkv-seg1@x</segment>
    </segments>
  </file>
</nzb>"#;
        let nzb = Nzb::parse(xml).expect("valid nzb");
        assert!(pick_canaries(&nzb).is_none());
    }

    #[test]
    fn sha1_short_is_16_hex_chars() {
        let s = sha1_short("https://example.com/foo.nzb|server1,server2");
        assert_eq!(s.len(), 16);
        assert!(s.chars().all(|c| c.is_ascii_hexdigit()));
    }
}
